### ISSUE TITLE
imagePullPolicy to pull latest image OR use existing image if repository unavailable

### ISSUE THREAD

**Issue by Alibirb** [kubernetes/kubernetes#111822]
### What would you like to be added?

Add another option for a container's `imagePullPolicy`, which attempts to pull the image, and falls back to using a local cached copy if the image could not be pulled.

### Why is this needed?

In development, images are frequently updated, so it's essential to grab the latest version of the image to test against.

In production, single points of failure should be minimized, so if the image is available through the repository OR cached locally, then the container should be able to start up (i.e. neither the repo nor the local cache is a single point of failure).

Currently, `imagePullPolicy` only offers three options:
* `Always`: Only tries to pull the image; ignores cache and fails if repository is unavailable
* `IfNotPresent`: First tries to use a cached image; falls back to pulling the image
* `Never`: Only tries to use cached image; ignores repository

From these options, `Always` is the only one that grabs the latest changes, so it's the only suitable option for development, but it turns the repository into a single point of failure, so it's not suitable for production unless you're okay with everything breaking if the repository goes offline.

In addition, using the `latest` tag (or any other tag that changes, like a `v1` tag that gives the latest 1.x.y version) only works if the `imagePullPolicy` is `Always`, or the image cache on each node is cleared whenever the tag is updated, so if those tags are used in production, and you don't want the repository to be a single point of failure, then there is NO suitable option for production at all.

A fourth option, which first tries to pull the image, and falls back to using a cached copy (the inverse of `IfNotPresent`), would be well-suited to both development and production, and would fit in very nicely in between `Always` and `IfNotPresent`.

Honestly, I think this would be the only `imagePullPolicy` I would ever use for ANYTHING I do (work and personal projects) if it were available, and I don't understand why it's missing.

**Comment by gjkim42** [comment]
+1
I think this feature is worth discussing.

/sig node

@bobbypage @pacoxu 
(sig node reviewers)

**Comment by pacoxu** [comment]
In our best practice guide
- user can use images' latest or v1 tag(that can change) with the Always pull policy in development env.
- - so every CD or restart of Pod can download the latest image.
- user should use a specific tag(that cannot change) of an image with `IfNotPresent` in production env. 
- - specific tags should not be updated in the image repository
- - specific tags can ensure that the image behavior can be predicted (Even if they are in a different node, the image should be the same for a specific tag.)

The fourth policy is like `PullOrUsePresent`. 
- the image can not be predicted. (It may be different in different nodes, as you are using a tag that may change.)

**Comment by JJotah** [comment]
Hi,
I'm agree with @pacoxu
The recommend for developers team is using tags. Kubernetes need to be a microservice and faster as posible. If anytime that you are creating the docker image the kubernetes need to pull the docker image will take some time.
I'm my opinion is better to follow tag strategy. If you want to pull always just put the image pull policy as always. But from my perspective I don't see this request as a optimal improvement.

**Comment by Alibirb** [comment]
I had really hoped we could avoid turning this into an argument about the validity of using mutable tags, but here goes.

According to [the Kubernetes website](https://kubernetes.io/docs/concepts/overview/what-is-kubernetes/#what-kubernetes-is-not), "Kubernetes aims to support an extremely diverse variety of workloads". I recognize that not everyone would have any use for this new `imagePullPolicy`, or mutable tags, but that's okay; nobody is going to force you to use it.

I know the typical recommendation is to use immutable tags so you can pin down the exact version you want to use, but that does NOT suit all use cases. For example, it may be desired to use a tag like `v1.2` which will update to the latest `1.2.x` version, so you automatically get bug fixes (especially security bug fixes) whenever a pod restarts. For a small business with a small IT team hosting their corporate website, file server, et cetera, it may be preferable to have automatic updates rather than have to constantly check for updates and manually change the configuration to a new version in order to keep the system secure.

And then there's situations like mine.

We were initially using immutable versioned tags, but we've switched to using `latest` except during development because we realized all we were doing by using immutable tags in testing and demo environments was forcing people to make more configuration changes, because **there was no use case for running an outdated version**.

The image I've been developing is just a monitoring sidecar, which gets added to pods created by over a dozen other companies, and deployed to a private cloud as part of a larger system-of-systems. The system-of-systems as a whole will eventually be installed on production hardware, formally tested (again), and then put into operation for multiple years, where it needs to keep working without system administrators to tend to it. There's no risk of outdated images in production, because the ONLY version of the image in the entire cloud is the version in the registry we delivered along with the rest of our subsystem. So in production, it doesn't matter whether we use immutable versioned tags, or `latest`, or even an SHA hash.

In testing or demos, it's a different story. Over a dozen companies need to add our sidecar to **their** pods. With immutable versioned tags, we'd have to tell them all to update their YAML every time we update our image, so it matches what's in the registry we give them as part of our next build. And of course, until they GET our next build (which is due at the same time as theirs is), that tag wouldn't work for them. And that's not even getting into what happens if compatibility issues forced someone to hold back the version of another subsystem in their integration testing environment. It was enough of a pain just going through this with our own test team. By just using `latest`, the pods will grab whatever version is in our registry, which is also the version that corresponds to the rest of our build they installed.

If, for some reason, the new version of our sidecar interferes with the rest of the pod, they can just remove the sidecar until we've fixed the compatibility issue, so there's no need to roll back to an earlier version. And an earlier version may not work with our latest build, anyway, because we may have made breaking API changes.

So, for us, using `latest` is what makes the most sense.

**Comment by Alibirb** [comment]
> * user should use a specific tag(that cannot change) of an image with `IfNotPresent` in production env.
> * * specific tags should not be updated in the image repository
> * * specific tags can ensure that the image behavior can be predicted (Even if they are in a different node, the image should be the same for a specific tag.)
> * Even some users use [alwayspullimages](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#alwayspullimages) admission control. They prefer to think that no images on nodes can be trusted.

If you're using immutable tags, then theoretically the image on the node is guaranteed to match what's in the registry. But you're saying there are still users who don't have full confidence that the local image hasn't been corrupted, or someone made a quick patch to the new release because they realized they tagged the wrong version or something. I could see that happening, and with a policy of `IfNotPresent`, if a node pulled the tag before it got corrected, the node would never update to the correct version of that tag. However, with `IfAvailable`, the next time the node starts a pod with that image, it would pull the corrected version, so the risk would be decreased in comparison to `IfAvailable`. You would need to push the wrong image, then stand up a pod that uses the new tag, push the correct image to the same tag, and then have the registry fail before the node stands up another pod with that tag, in order to wind up running a version that doesn't match what was in the registry. Still a risk, sure, but using `Always` could be considered riskier since your pod flat-out wouldn't run if the registry became unavailable, so `IfAvailable` would allow to achieve a balance between resiliency to registry issues and ensuring you run the right version.

**Comment by pacoxu** [comment]
It makes sense in your auto-upgrade cases. 

However, it sounds like the image pulling sometimes failed in your env. (Is this the key reason for adding this policy?) I'd like to know why there are registry failures. 
- `IfAvailable` is like  `BestEffortAlways` that allow image pull failure. (It seems to be a strategy for edge env whose network is not stable.)

**Comment by JJotah** [comment]
Hi,
I'm sorry. I understand your point but this feature works for local environments to test the image. Dev environment, staging a production if you have some network issues you can't test the service or the image outside the cluster. Anyway understanding your point is better to add new functionality into image pull policy like imagePullPolicyconnectionError true. Or something like that.

**Comment by Alibirb** [comment]
Our registry is fairly reliable, but everything has to be really fault-tolerant, because if something stops working in production, the users need to be able to either quickly fix it by rebooting (before the unavailability causes much bigger issues), or carry out their duties without it. So having our registry be a single point of failure for everything in the cluster is a tough sell, and would become even more difficult if any subsystem ever moves one of their safety-critical components into the cluster.

So basically, it's a case of "take all appropriate measures", and letting Kubernetes fall back to using a local image if the registry is unavailable seems like a common sense measure to take.

**Comment by droctothorpe** [comment]
+1. I'd be interested in taking a stab at this potentially. Should be pretty easy to implement. Just don't want to do the work if the issue isn't signed off on.

**Comment by Alibirb** [comment]
> Hi, I'm sorry. I understand your point but this feature works for local environments to test the image. Dev environment, staging a production if you have some network issues you can't test the service or the image outside the cluster. Anyway understanding your point is better to add new functionality into image pull policy like imagePullPolicyconnectionError true. Or something like that.

(Sorry for the late response; somehow I didn't notice your comment before.)

Just because the registry isn't available, doesn't mean that all network connectivity is gone. It's possible that the service running the registry has failed while other services remain functional, so if we could stand up a container with the image, it would still work. We also have another container we'll be adding soon, which would be able to continue to gather data even if network connectivity was completely gone, and then send the data when connectivity is restored. So having `imagePullPolicy: IfAvailable` would be helpful for cases like that.

Could you explain what you mean by `imagePullPolicyconnectionError true`?

**Comment by Alibirb** [comment]
I'd be interested in implementing this myself if we can reach some sort of tentative agreement to accept the feature. I'm honestly amazed Kubernetes has gone so long without this, as it seems like the ideal behavior for any use case where operational availability is a greater concern than potentially outdated images.

/remove-lifecycle-stale

**Comment by gjkim42** [comment]
@Alibirb 

You could propose the feature and get some more feedback at the [sig-node meeting](https://github.com/kubernetes/community/blob/master/sig-node/README.md) before starting.

**Comment by Alibirb** [comment]
> Am I missing something? Is there another solution out there?

Looks like someone linked to this issue in kube-image-keeper, which is a proxy to locally cache images from remote registries. It doesn't support mutable tags but it sounds like they're working on it. If they get that working, it would provide *somewhat* of a workaround for this issue in some cases, by making the local caching proxy the point of failure instead of the remote registry. Whether that's a better or a worse point of failure would depend on the reliability/stability of kube-image-keeper, and the specific use case (i.e. can you tolerate one or two nodes that can't start up containers but k8s still tries to assign to them?). In general it's probably more reliable, as it protects from network disruption, but it increases the complexity of the cluster, and requires a mutating webhook and a local proxy service on each node. If that local proxy fails, then any container on that node that uses `Always` would fail to start up.

`IfAvailable` would be a far cleaner solution because it uses the existing cache on each node instead of introducing an additional cache layer.

**Comment by Alibirb** [comment]
Another work-around in some cases might be to create a mutating webhook that replaces mutable tags and `Always` with the latest hash for that tag at the time of pod creation and changes the policy to `IfNotPresent`. This would achieve a similar effect as `IfAvailable`, but could still fail in some circumstances, e.g. if the webhook can reach the registry but the node the pod is assigned to can't (due to network issues, credentials, rate limiting, or poor timing), and the image for that hash is not cached. It also would result in an imagePullPolicy of `IfNotPresent`, which it sounds like is against the CIS hardening so if you're constrained to that, then it doesn't help. And, of course, something could go wrong with the webhook.

Overall, it seems like there are workarounds for different scenarios, but nothing as clean and reliable as simply allowing the node to fall back to the version it already has cached.

**Comment by StCyr** [comment]
@alibirb,

is there something that prevents you to make a PR?

May I help?

**Comment by Alibirb** [comment]
Mostly, I just wanted some kind of agreement that this was something that might actually get accepted in before I devoted the time to making the change.

I keep meaning to attend the SIG node meeting and bring it up, but then something comes up at work at the time.

**Comment by Kahz3l** [comment]
I would also love to have this feature.

**Comment by LufinityLucas** [comment]
+1 on this feature. If you have a private registry inside the K8S environment (e.g., a Gitea container) that craches due to a node crach, other containers won't be able to spin back up since the registry is down. The registry might come up eventually, but let's say the registry container depends on a DB container that can't start either, since the registry is unreachable. Moment 22... Here, an option to always pull *if possible*, otherwise use a cached image, would be really helpful.

**Comment by dbowling** [comment]
Documenting other relevant use cases:

Last year, Broadcom deleted their vSphere CSI drive image registry before having a new registry populated with all the supported CSI images. The vendor's install had used `imagePullPolicy: Always`, presumably due to mutability. Since a CSI driver is a critical cluster component, when we restarted a node, it failed to attach any PVCs. If another option (if possible or the like) was added, vendors might be more likely to use that if they feel they wish to force constant re-pulls on us.

We've been bitten by throttling from some registries as well, mostly in the lower dev environments. The current configuration options have us choose between extremes, when we would prefer the middle ground.

The workaround of never setting `Always`, and instead using `IfNotPresent` or `Never` with a custom solution to provide images (e.g. something like [kubernetes-image-puller](https://github.com/che-incubator/kubernetes-image-puller)) is viable, but does require substantially more work to implement/maintain (from a cluster admin perspective) than what is proposed here.

**Comment by Kahz3l** [comment]
/remove-lifecycle stale
This is an important topic to many of us

**Comment by SergeyKanzhelev** [comment]
The problem with this feature is that it promotes the bad practice of using labels in production. If we are ignoring this for a moment, let's clarify the requirements here. These are requirements listed in the issue description:

1. Everything is working: Try resolve the tag. Download latest or use one on a disk.
2. Registry is down: Try to resolve the tag. Failing to resolve tag. Return the image from the disk.

What would be the expected behavior in these edge case - Registry crashed after resolving the tag, let's say `latest` (or resolved to non-existing version). So `kubelet` called `PullImage`. And runtime knows that the `latest` is no longer pointing to the version on the disk (successful call to get the metadata). Should runtime remove the `latest` from the on-disk version? I would say yes. Or it must preserve the tag on the on-disk version, which will be inconsistent behavior? I do not think we have this behavior documented. So there is a good chance that `latest` is not pointing to the on-disk version of an image after `PullImage` existed with the error. So `GetImage` will not return anything. Even if kubelet will call `GetImage` before the `PullImage`, the next attempt to use it will fail. So we have not really protected workload from the registry crash.

Also, when using image pull secrets, it is expected that on bootstrap some secretes may not be available for the first pull attempt. If this error to download will result in using the cached version, it may be a very unexpected behavior. Let's say the images preloaded on a node are very old. Nodes are created from some "template" with these old preloaded images, and every time on bootstrap there will be old image used for the first container.

So implementing this behavior in straightforward way may not protect from all registry failures, and may result in unexpected behavior in edge cases. Thus, more designing effort is needed and since the issue promoting the bad use of labels in production, it is a low priority request.

Whoever wants to pick up this task, a good starting point will be a full writeup on the change, edge cases, and review it with runtime maintainers (cri-o and containerd) as I believe it will require to change cri api.

**Comment by SergeyKanzhelev** [comment]
> /remove-lifecycle stale

@Kahz3l is removing of a stale represents the fact that it is still desired feature for you?

**Comment by Kahz3l** [comment]
Hi Sergey,
Yes it is still a desirable feature even with the arguments that you presented.
It sounds like it would be a lot of work and yes I agree that it would not be best practice.
But in development systems I don't care that much about it not being best practice, I just want to test the newest version and that would simplify it a lot.
In my production system I obviously would pin it to an immutable tag or digest.

It's like how a workshop mechanic obviously knows he should use a seat belt when driving across the workshop parking lot but never does because it is inconvenient and the chances of something happening is small.

**Comment by Starttoaster** [comment]
Somewhat bad analogy. Assuming its use is in development or testing environments, it would be more like if the workshop mechanic was driving a remote controlled car across the workshop parking lot to see if they're going to have unforeseen trouble when they drive across it for real.
